### DISCUSSION TITLE
[Feature] Link Faces with Contacts

### DISCUSSION THREAD

**Discussion by palitu** [immich-app/immich#2629] [Feature Request]
### The feature

**Associate Faces with a contact**

One thing that i would love to see is a way to join the matched faces in Immich with my contacts list.  It would be good to be able to join to my CardDAV (radicale).

One of the issues that i always have is determining who to name, and who not to name.  Similar to my birthdays list & Addresses, how do i keep all those sort of details together.

I expect that it should be relatively easy.

1. Connect to CalDav server (or other contact provider)
2. associate an immich named user with a contact

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by NicholasFlamy** [comment]
At the bare minimum, a feature in the mobile app to auto fill from contacts would be great. Google Photos and Samsung Gallery both have it. They also run Face recognition on the contact photo to try and guess what contact is what person but that isn't necessary.

**Reply by NicholasFlamy** [reply]
![Screenshot_20240202_160457_Gallery](https://github.com/immich-app/immich/assets/30300649/3092d034-4988-4e9c-bfef-eb9fba659879)
![Screenshot_20240202_160441_Gallery](https://github.com/immich-app/immich/assets/30300649/e63b4e00-d03c-4187-ae64-187160f40cf9)
![Screenshot_20240202_160323_Photos](https://github.com/immich-app/immich/assets/30300649/0e1048f1-86ed-42e9-94a7-568b8dd927e7)
![Screenshot_20240202_160302_Photos](https://github.com/immich-app/immich/assets/30300649/659e9cd2-6f9a-47fa-bcfa-83c37881d47c)

**Reply by NicholasFlamy** [reply]
There was a PR that implemented this but the main issue was the way the app asked for the Contacts permission.

**Comment by craigS89** [comment]
This would be a great feature to have.

**Comment by axgd-code** [comment]
think this feature could be useful for several other upcoming functionalities:

- Preparing albums or memories related to the person's birthday,
- Synchronizing an Immich link to the user's contact sheet.
- Choosing a photo to synchronize with the user's contact sheet.

**Comment by minecrawler** [comment]
Missing bridges to other software via standard protocols is the reasons I'm not using immich. I want it to share data, so I only have to manage everything once. One big example is CardDAV. Immich is not a people management app, so I want it to pull contact info from another application and use that to connect contacts to faces. The other way around I want to ask immich which photos and videos are connected to one of my contacts and then get that information back; or ask what assets are connected to a tag or search term. immich should focus on what it's good at: photos and videos. And use APIs to connect to everything else to have other specialized applications doing what they do best,

For example, NextCloud could be a DAV provider, and SearXNG could be used as a central search which pulls data from immich.

**Comment by DasIgeli** [comment]
I think the easiest way would be to utilize the mobile apps which write the contact information to the Database if you decide to link it. This way you have a standard protocol on the major phone OS, which can be utilized and then this info is stored in a separate table. 

Upon linking a person to a photo the dropdown menu will show all your contacts and this will contain the birthday information. Would be super nice!


Editing, so that I can remove the other topic, I've opened.
I love to see how old people were at given points in time, when photos were taken, but it's a taxing to fill this information for each (new) person despite having it with me, on my phone.

Thus I'd like to officially request the import of the phones address list (or any vCard-type-list) into immich, which can then be used to link people and enrich the information, such as birthdays for later use.

Benefit: Many of us likely have all the important information ready in our address lists on our phones.
After the import of the contacts you could link them to people within immich and fill information such as Name, Birthday, Favorite-Status automatically without having users to manually enter this again.

I feel like there would be a great benefit in having such a feature.

[Apple - Accessing Contact Store](https://developer.apple.com/documentation/contacts/accessing-the-contact-store)

[Android - Contacts Provider](https://developer.android.com/identity/providers/contacts-provider)

**Comment by lwctr** [comment]
This feature seems like one of the bigger differences between something like Google photos and immich. In general enabling immich to incorporate into a foss (or even not foss) ecosystem of different apps (in this case just a contact provider) would greatly improve the appeal of immich.

I also think that allowing the user to choose the source of their contact information would be rather nice, as not every person using immich also uses WebDAV or some other thing to sync their contacts, maybe they just want to have immich link to contacts on their phone. On the other hand some users might not want to link all their contacts to faces in their gallery.

**Comment by CraftoHohenvels** [comment]
Since we already have an option to add an birthdate to people/faces, this should also be synced with the contacts. I don't see a reason to enter the birthdate of a person twice in two different apps.

**Reply by daniele-athome** [reply]
You could use my tool to sync birth dates from a CardDAV server: https://github.com/daniele-athome/immich-carddav-sync

Personally I believe this (I mean the "syncing with contacts" part) is something that doesn't really belong in Immich and can (and should) be provided by 3rd party tools.

**Reply by NicholasFlamy** [reply]
I do think that it's pretty cool on the server side.

**Reply by CraftoHohenvels** [reply]
I personally think it's absolutely something immich should do.
Users seeking for alternatives to Google Photos usually want something that is very similar and is convenient for them.
In other words if they install immich, find that feature xy is not supported and then have to look for a tool to add this feature (in a likely inconvenient way) and then have to tinker to get that working, they will probably be sad that this feature does not exist and/or use a different platform or even go back to Google Photos, as it is just "more convenient".

I am a software engineer myself and from experience I can confidently say users don't want to change their habits and usually want it as simple (non technical) as possible. A solution with a 3rd party tool is neither.

I also want to mention that I don't see this as a CardDAV exclusive feature, because if the user migrated from Google Photos (or Apple Photos) it is likely they also use the contacts app that came with their phone (e.g. Google Contacts or Apple Contacts)

**Reply by NicholasFlamy** [reply]
> I also want to mention that I don't see this as a CardDAV exclusive feature, because if the user migrated from Google Photos (or Apple Photos) it is likely they also use the contacts app that came with their phone (e.g. Google Contacts or Apple Contacts)

There was a working PR for this in the app, but the main issue is it adds a permission and it wasn't handling how that permission is dealt with very well. I don't think the devs want this in the app.

Additionally for server side, I absolutely agree that Immich itself should NOT do it and that a third party tool should. Unlike Google Photos, there is no Immich suite. Maybe in Nextcloud photos it'd be built in.

The solution is someone should just help by making a Docker compose that includes the third party tool and it just works pretty easy.

**Reply by CraftoHohenvels** [reply]
For server side I would agree.
I personally see this as a feature for the mobile app too.

> The solution is someone should just help by making a Docker compose that includes the third party tool and it just works pretty easy.

Fair point. Though I still see the kind of user I mentioned looking on the official site. I personally don't see such a compose being added there.

I also just found the PR https://github.com/immich-app/immich/pull/15817 
Looking through it I don't see any feature from this discussion/Feature Request being added, mainly because of the reasoning "It requires an extra permission that feels intrusive to user privacy".

**Reply by daniele-athome** [reply]
> I am a software engineer myself and from experience I can confidently say users don't want to change their habits and usually want it as simple (non technical) as possible. A solution with a 3rd party tool is neither.

I totally agree about habits etc. I also think it depends on the 3rd party tool. If it's easy to install (e.g. "add this block of text to your docker-compose.yml") and provides an easy way to set it up somehow (e.g. a UI), it could convince a significant number of users to adopt it. Besides, running Immich itself requires at least a tiny bit of tinkering anyway - we are talking about self-hosting: a bare minimum of system administration knowledge is required.

**Reply by boul2gom** [reply]
> > I am a software engineer myself and from experience I can confidently say users don't want to change their habits and usually want it as simple (non technical) as possible. A solution with a 3rd party tool is neither.
> 
> I totally agree about habits etc. I also think it depends on the 3rd party tool. If it's easy to install (e.g. "add this block of text to your docker-compose.yml") and provides an easy way to set it up somehow (e.g. a UI), it could convince a significant number of users to adopt it. Besides, running Immich itself requires at least a tiny bit of tinkering anyway - we are talking about self-hosting: a bare minimum of system administration knowledge is required.

Doing this on mobile is tricky because of iOS/Android permissions...A separate app or plugins doesn't solve the privacy issue. Immich (or the second app) would still need local contact permissions, OS sandboxing prevents plugins from bypassing this, and it adds Store review headaches.
And a server-side Docker container works great with CardDAV, but it's completely blind to contacts stored locally on a phone unless you manually export a .vcf file.
But on iOS, permissions aren't asked to user only when trying to access a specific feature that require them ? It may solve the privacy issue, because contacts access is only granted if this feature is used no ?

**Reply by NicholasFlamy** [reply]
> A separate app or plugins doesn't solve the privacy issue. Immich (or the second app) would still need local contact permissions, OS sandboxing prevents plugins from bypassing this, and it adds Store review headaches.

The idea would be to only include this feature on the web and leave it out or use the server's connected CardDAV server for a source. So it does indeed solve the issue.



> And a server-side Docker container works great with CardDAV, but it's completely blind to contacts stored locally on a phone unless you manually export a .vcf file.

I hope for any user's sake that user is at least syncing to somewhere such as their own server or Google Contacts so they don't lose their contacts (unless they're all replaceable).

**Reply by MrRulf** [reply]
Just found all of this. I agree it would be a great feature to have the photo recognition pull info from contacts, auto recognise from their contact picture and sync to check for changes (corrections, whatever). 

Correct me if I'm wrong but I think the concept is already quite clear on how this feature should work and be implemented. I added some thoughts from myself:

* Solved through carddav sync on the server
  * Maybe also other protocols like jmap
  * Maybe read-only
* Immich should not turn into a monolith that now also does contacts
  * Immich is a carddav client
  * A discussion is needed on how much immich needs to store/cache/manage locally
    * Immich should have as little stuff as necessary locally to keep the single source of truth in carddav
    * Immich might need to have some mirror or cache to connect contacts and picture groups without bloating the contacts in the cardDAV server. I'm not familiar with cardDAV, but seeing that they are cached in thunderbird and davx5 anyway this might already be solved through that.
  * Maybe this can be realised modular through an additional component like done with the machine learning, if practical. But this should be an official feature, not some random thing from a random third party. Machine learning is also it's own container and that's fine for setup so it would be fair to assume this should also be no problem for the admins.
* It'd be good if the contact sync can be setup per account and not globally for all users. Some might have Google, some nextcloud, some radicale, some Apple. 
* Having this solved through the client app and contacts there (apple/android) is secondary, might not be realised at all because of permission problems and potential conflicts with the server client for contacts. Not a focus for now. 

Is there anything I forgot or does this sum it up quite well until now? Is there something to change still?

**Reply by daniele-athome** [reply]
If you look around, there are already multiple implementations of this (besides mine) that fits well as a sidecar container. There's nothing inherently wrong with having 3rd parties improving the ecosystem around Immich.

As for the technical issues of this: it's not that easy because not all providers expose CardDAV APIs (and CardDAV itself is a mess already).

**Reply by MrRulf** [reply]
The issue is not that third parties improve the ecosystem, but that this would be a great native feature. The arguments for that have already been brought and to put it this way, I don't really see any arguments why this shouldn't be a native feature. Having it native the internal name+birthday+face system could also be better and directly integrated with it, while an external tool pulling from one system and pushing into another leaves quite a bit more room for problems. 

CardDAV seems stable enough, i mean it works in enough places across enough providers and clients and I think at least the big players offer cardDAV endpoints, so while there probably are contact syncs that don't have carddav in any form, at least 90% or 95% should be fine and you gotta start somewhere

**Reply by MrRulf** [reply]
also third party tools bring the limitation of mostly being dependent on the admin to be configured for each user
